src/features/hyperlinks.rs

```diff
@@ -19,26 +19,83 @@ pub fn make_feature() -> Vec<(String, OptionValueFunction)> {
+}
+
+#[cfg(not(test))]
+pub fn remote_from_config(cfg: &Option<&GitConfig>) -> Option<GitRemoteRepo> {
+    cfg.and_then(GitConfig::get_remote_url)
+}
+
 pub fn format_commit_line_with_osc8_commit_hyperlink<'a>(
     line: &'a str,
     config: &Config,
 ) -> Cow<'a, str> {
+    // Given a [START..END]-match and pos = 0, store [pos..START] first, then the hyperlink(match),
+    // then store pos = END. Repeat while matches exist, and at the end store [pos..].
     if let Some(commit_link_format) = &config.hyperlinks_commit_link_format {
-        COMMIT_LINE_REGEX.replace(line, |captures: &Captures| {
-            let prefix = captures.get(1).map(|m| m.as_str()).unwrap_or("");
-            let commit = captures.get(2).map(|m| m.as_str()).unwrap();
-            let suffix = captures.get(3).map(|m| m.as_str()).unwrap_or("");
-            let formatted_commit =
-                format_osc8_hyperlink(&commit_link_format.replace("{commit}", commit), commit);
-            format!("{prefix}{formatted_commit}{suffix}")
-        })
-    } else if let Some(repo) = config.git_config().and_then(GitConfig::get_remote_url) {
-        COMMIT_LINE_REGEX.replace(line, |captures: &Captures| {
-            format_commit_line_captures_with_osc8_commit_hyperlink(captures, &repo)
-        })
-    } else {
-        Cow::from(line)
+        let mut matches = COMMIT_LINE_REGEX.find_iter(line);
+        if let Some(capture) = matches.next() {
+            let mut pos = 0;
+            let mut result = String::new();
+
+            let mut create_commit_hyperlink = |pre_cap_pos, cap: Match| {
+                result.push_str(&line[pre_cap_pos..cap.start()]);
+                let commit = &line[cap.start()..cap.end()];
+                let formatted_commit =
+                    format_osc8_hyperlink(&commit_link_format.replace("{commit}", commit), commit);
+                result.push_str(&formatted_commit);
+                cap.end()
+            };
+
+            pos = create_commit_hyperlink(pos, capture);
+
+            for capture in matches {
+                pos = create_commit_hyperlink(pos, capture);
+            }
+
+            result.push_str(&line[pos..]);
+            return Cow::from(result);
+        }
+    } else if let Some(repo) = remote_from_config(&config.git_config()) {
+        let mut matches = COMMIT_LINE_REGEX.find_iter(line);
+        if let Some(capture) = matches.next() {
+            let mut pos = 0;
+            let mut result = String::new();
+
+            let mut create_repo_hyperlink = |pre_cap_pos, cap: Match| {
+                result.push_str(&line[pre_cap_pos..cap.start()]);
+                let commit = &line[cap.start()..cap.end()];
+                let formatted_commit = format!(
+                    "{osc}8;;{url}{st}{commit}{osc}8;;{st}",
+                    url = repo.format_commit_url(commit),
+                    commit = commit,
+                    osc = "\x1b]",
+                    st = "\x1b\\"
+                );
```

Comment:
Is there a reason that we can't use `format_osc8_hyperlink` here also? And beyond that, The two branches of this function are quite substantial and look very similar -- did you decide it wasn't attractive to unite them?

Reply:
Right, just didn't want think about how to nicely factor that out :) - What about this `HyperlinkCommits` helper now?

Reply:
Nice, the helper LGTM.